src/relay/ir/dataflow_matcher.cc

```diff
@@ -0,0 +1,421 @@
+                  auto mul =
+                      CallPatternNode::make(op->op, {arg_node->args[0], op->args[(arg_id + 1) % 2]},
+                                            op->attrs, op->type_args);
+                  auto div = CallPatternNode::make(arg_node->op, {mul, arg_node->args[1]},
+                                                   arg_node->attrs, arg_node->type_args);
+                  return VisitDFPattern(div, expr);
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  return false;
+}
+
+bool DFPatternMatcher::VisitDFPattern_(const DominatorPatternNode* op, const Expr& expr) {
+  auto watermark = matched_nodes_.size();
+  auto backup_memo = memo_;
+  auto backup_matched_nodes = matched_nodes_;
+
+  if (VisitDFPattern(op->child, expr)) {
+    auto child_graph = CreateIndexedGraph(GetRef<DFPattern>(op));
+    auto expr_graph = CreateIndexedGraph(expr);
+    auto find_dominated = [&child_graph, this](const DFPattern& node) {
+      std::unordered_set<Expr, ObjectHash, ObjectEqual> dominated_exprs;
+      auto indexed_node = child_graph.node_map_[node];
+      for (auto dominated : indexed_node->dominator_children_) {
+        if (dominated->ref_.as<WildcardPatternNode>() || dominated->ref_.as<OpNode>()) {
+          continue;
+        }
+        dominated_exprs.insert(memo_[dominated->ref_]);
+      }
+      return dominated_exprs;
+    };
+    std::function<bool(const Expr&, const std::unordered_set<Expr, ObjectHash, ObjectEqual>&)>
+        find_parent;
+    find_parent = [this, &op, &watermark, &backup_memo, &backup_matched_nodes, &find_dominated,
+                   &expr_graph, &find_parent](
+                      const Expr& expr,
+                      const std::unordered_set<Expr, ObjectHash, ObjectEqual>& dominated_exprs) {
+      bool out = true;
+      for (auto node : expr_graph.node_map_[expr]->dominator_children_) {
+        if (out && dominated_exprs.count(node->ref_) == 0) {
+          if (VisitDFPattern(op->parent, node->ref_)) {
+            backup_memo[op->parent] = memo_.at(op->parent);
+            backup_matched_nodes.push_back(op->parent);
+            memo_ = backup_memo;
+            matched_nodes_ = backup_matched_nodes;
+            watermark += 1;
+            return true;
+          } else {
+            if (VisitDFPattern(op->path, node->ref_)) {
+              auto new_dominated_exprs = find_dominated(op->path);
+              ClearMap(watermark);
+              out &= find_parent(node->ref_, new_dominated_exprs);
+            } else {
+              out = false;
+            }
+          }
+        }
+      }
+      return out;
+    };
+
+    auto dominated_exprs = find_dominated(op->child);
+    ClearMap(watermark);
+    bool matches = find_parent(expr, dominated_exprs);
+    if (matches) {
+      backup_memo[op->parent] = memo_.at(op->parent);
+      backup_memo[op->child] = expr;
+      memo_ = backup_memo;
```

Comment:
why not directly update `memo_`?

Reply:
:/ this is kind of leftover from an earlier design, I guess there's no real reason anymore. Will simplify, thanks.